client/service/src/chain_ops.rs

```diff
@@ -36,12 +37,116 @@ use sc_executor::{NativeExecutor, NativeExecutionDispatch};
 
 use std::{io::{Read, Write, Seek}, pin::Pin};
 use sc_client_api::BlockBackend;
+use serde_json::{de::IoRead as JsonIoRead, Deserializer, StreamDeserializer};
 
 /// Build a chain spec json
 pub fn build_spec(spec: &dyn ChainSpec, raw: bool) -> error::Result<String> {
 	Ok(spec.as_json(raw)?)
 }
 
+/// Helper enum that wraps either a binary decoder (from parity-scale-codec), or a JSON decoder (from serde_json).
+enum BlockStream<R, B> 
+where
+	R: std::io::Read + std::io::Seek,
+{
+	Binary {
+		count: u64,
+		read_block_count: u64,
```

Comment:
wow, I was wondering what this was for until I saw its use in the `Stream` impl, maybe rename this or add docs. Actually now that i think about it, docs would be preferable. :pray: 
```suggestion
		num_read_blocks: u64,
```

Reply:
I've added some comments now, let me know what you think about it! Keen to hear your feedback :)